middle_end/flambda2/reaper/traverse.ml

```diff
@@ -0,0 +1,672 @@
+      in
+      let group = List.rev rev_group in
+      Static_consts group
+    (* TODO set_of_closures in Static_consts *)
+    | Prim _ -> Named defining_expr
+    | Simple _ -> Named defining_expr
+    | Rec_info _ as defining_expr -> Named defining_expr
+  in
+  let let_acc =
+    Let { bound_pattern; defining_expr = named; parent = denv.parent }
+  in
+  traverse
+    { parent = let_acc;
+      conts = denv.conts;
+      current_code_id = denv.current_code_id
+    }
+    acc body
+
+and traverse_prim denv acc ~bound_pattern (prim : Flambda_primitive.t) ~default
+    ~(default_bp : acc -> Graph.Dep.t -> unit) =
+  let known_field_of_block field block =
+    Simple.pattern_match field
+      ~name:(fun _ ~coercion:_ -> None)
+      ~const:(fun cst ->
+        Simple.pattern_match block
+          ~const:(fun _ -> None)
+            (* CR ncourant: it seems this const case can happen with the
+             * following code:
+             *
+             * let[@inline] f b x = if b then Lazy.force x else 0
+             * let g b = f b (lazy 0)
+             *
+             * It is unclear why it has not been transformed by an Invalid by
+             * simplify, however.
+             *)
+          ~name:(fun block ~coercion:_ ->
+            match[@ocaml.warning "-4"] Int_ids.Const.descr cst with
+            | Tagged_immediate i ->
+              let i = Targetint_31_63.to_int_exn i in
+              Some (i, block)
+            | _ -> assert false))
+  in
+  let () =
+    let kind = Flambda_primitive.result_kind' prim in
+    let name =
+      Name.var (Bound_var.var (Bound_pattern.must_be_singleton bound_pattern))
+    in
+    Acc.kind name kind acc
+  in
+  match[@ocaml.warning "-4"] prim with
+  | Variadic (Make_block (_, _mutability, _), fields) ->
+    List.iteri
+      (fun i field ->
+        Simple.pattern_match field
+          ~name:(fun name ~coercion:_ ->
+            default_bp acc
+              (Block { relation = Block i; target = Code_id_or_name.name name }))
+          ~const:(fun _ -> ()))
+      fields
+  | Unary (Project_function_slot { move_from = _; move_to }, block) ->
+    let block =
+      Simple.pattern_match block
+        ~name:(fun name ~coercion:_ -> name)
+        ~const:(fun _ -> assert false)
+    in
+    default_bp acc (Field { relation = Function_slot move_to; target = block })
+  | Unary (Project_value_slot { project_from = _; value_slot }, block) ->
+    let block =
+      Simple.pattern_match block
+        ~name:(fun name ~coercion:_ -> name)
+        ~const:(fun _ -> assert false)
+    in
+    default_bp acc (Field { relation = Value_slot value_slot; target = block })
+  | Binary (Block_load (_access_kind, _mutability), block, field) -> (
+    (* Loads from mutable blocks are tracked here. This is ok as long as stores
+       are properly tracked also. This is a flow-insensitive dependency
+       analysis: this might produce surprising results sometimes. *)
```

Comment:
Reword comment: this is not the case anymore since we can't track block_set properly